File: cpp/c-api/src/transport/serialization/json/serializable_document.cpp
#include "dnv/vista/sdk/c/transport/serialization/json/serializable_document.h"

#include "../../../cast_internal.h"
#include "../../../error_internal.h"

#include <memory>

using namespace dnv::vista::sdk;
using dnv::vista::sdk::c::fromSerializableDocument;
using dnv::vista::sdk::c::fromSerializableDocumentRef;
using dnv::vista::sdk::c::toSerializableDocument;

using Doc = transport::serialization::json::SerializableDocument;

static_assert(static_cast<int>(Doc::Kind::Null) == DNV_VISTA_SDK_SERIALIZABLE_DOCUMENT_KIND_NULL);
static_assert(static_cast<int>(Doc::Kind::Boolean) == DNV_VISTA_SDK_SERIALIZABLE_DOCUMENT_KIND_BOOLEAN);
static_assert(static_cast<int>(Doc::Kind::Integer) == DNV_VISTA_SDK_SERIALIZABLE_DOCUMENT_KIND_INTEGER);
static_assert(static_cast<int>(Doc::Kind::Double) == DNV_VISTA_SDK_SERIALIZABLE_DOCUMENT_KIND_DOUBLE);
static_assert(static_cast<int>(Doc::Kind::String) == DNV_VISTA_SDK_SERIALIZABLE_DOCUMENT_KIND_STRING);
static_assert(static_cast<int>(Doc::Kind::Array) == DNV_VISTA_SDK_SERIALIZABLE_DOCUMENT_KIND_ARRAY);
static_assert(static_cast<int>(Doc::Kind::Object) == DNV_VISTA_SDK_SERIALIZABLE_DOCUMENT_KIND_OBJECT);

dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_null(void)
{
    return fromSerializableDocument(Doc{});
}

dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_from_boolean(int value)
{
    return fromSerializableDocument(Doc{ value != 0 });
}

dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_from_integer(int64_t value)
{
    return fromSerializableDocument(Doc{ value });
}

dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_from_double(double value)
{
    return fromSerializableDocument(Doc{ value });
}

dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_from_string(const char* value)
{
    if (value == nullptr)
    {
        c::setLastErrorMessage("value must not be null");
        return nullptr;
    }

    return fromSerializableDocument(Doc{ std::string{ value } });
}

dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_array(void)
{
    return fromSerializableDocument(Doc::array());
}

dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_object(void)
{
    return fromSerializableDocument(Doc::object());
}

void dnv_vista_sdk_serializable_document_free(dnv_vista_sdk_serializable_document_t* doc)
{
    delete reinterpret_cast<Doc*>(doc);
}

dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_clone(
    const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return nullptr;
    }

    return fromSerializableDocument(Doc{ *toSerializableDocument(doc) });
}

int dnv_vista_sdk_serializable_document_equals(
    const dnv_vista_sdk_serializable_document_t* a, const dnv_vista_sdk_serializable_document_t* b)
{
    if (a == nullptr || b == nullptr)
    {
        c::setLastErrorMessage("a and b must not be null");
        return 0;
    }

    return *toSerializableDocument(a) == *toSerializableDocument(b) ? 1 : 0;
}

dnv_vista_sdk_serializable_document_kind_t dnv_vista_sdk_serializable_document_kind(
    const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return DNV_VISTA_SDK_SERIALIZABLE_DOCUMENT_KIND_NULL;
    }

    return static_cast<dnv_vista_sdk_serializable_document_kind_t>(toSerializableDocument(doc)->kind());
}

int dnv_vista_sdk_serializable_document_is_null(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0;
    }

    return toSerializableDocument(doc)->isNull() ? 1 : 0;
}

int dnv_vista_sdk_serializable_document_is_boolean(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0;
    }

    return toSerializableDocument(doc)->isBoolean() ? 1 : 0;
}

int dnv_vista_sdk_serializable_document_is_integer(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0;
    }

    return toSerializableDocument(doc)->isInteger() ? 1 : 0;
}

int dnv_vista_sdk_serializable_document_is_double(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0;
    }

    return toSerializableDocument(doc)->isDouble() ? 1 : 0;
}

int dnv_vista_sdk_serializable_document_is_string(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0;
    }

    return toSerializableDocument(doc)->isString() ? 1 : 0;
}

int dnv_vista_sdk_serializable_document_is_array(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0;
    }

    return toSerializableDocument(doc)->isArray() ? 1 : 0;
}

int dnv_vista_sdk_serializable_document_is_object(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0;
    }

    return toSerializableDocument(doc)->isObject() ? 1 : 0;
}

int dnv_vista_sdk_serializable_document_as_boolean(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0;
    }

    try
    {
        return toSerializableDocument(doc)->asBoolean() ? 1 : 0;
    }
    catch (const std::exception& e)
    {
        c::setLastErrorMessage(e.what());
        return 0;
    }
}

int64_t dnv_vista_sdk_serializable_document_as_integer(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0;
    }

    try
    {
        return toSerializableDocument(doc)->asInteger();
    }
    catch (const std::exception& e)
    {
        c::setLastErrorMessage(e.what());
        return 0;
    }
}

double dnv_vista_sdk_serializable_document_as_double(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return 0.0;
    }

    try
    {
        return toSerializableDocument(doc)->asDouble();
    }
    catch (const std::exception& e)
    {
        c::setLastErrorMessage(e.what());
        return 0.0;
    }
}

const char* dnv_vista_sdk_serializable_document_as_string(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr)
    {
        c::setLastErrorMessage("doc must not be null");
        return nullptr;
    }

    try
    {
        return toSerializableDocument(doc)->asString().c_str();
    }
    catch (const std::exception& e)
    {
        c::setLastErrorMessage(e.what());
        return nullptr;
    }
}

size_t dnv_vista_sdk_serializable_document_array_size(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr || !toSerializableDocument(doc)->isArray())
    {
        return 0;
    }

    return toSerializableDocument(doc)->asArray().size();
}

const dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_array_at(
    const dnv_vista_sdk_serializable_document_t* doc, size_t index)
{
    if (doc == nullptr || !toSerializableDocument(doc)->isArray())
    {
        c::setLastErrorMessage("doc must not be null and must be an array");
        return nullptr;
    }

    const auto& arr = toSerializableDocument(doc)->asArray();
    if (index >= arr.size())
    {
        c::setLastErrorMessage("index out of range");
        return nullptr;
    }

    return fromSerializableDocumentRef(arr[index]);
}

size_t dnv_vista_sdk_serializable_document_object_size(const dnv_vista_sdk_serializable_document_t* doc)
{
    if (doc == nullptr || !toSerializableDocument(doc)->isObject())
    {
        return 0;
    }

    return toSerializableDocument(doc)->asObject().size();
}

const char* dnv_vista_sdk_serializable_document_object_key_at(
    const dnv_vista_sdk_serializable_document_t* doc, size_t index)
{
    if (doc == nullptr || !toSerializableDocument(doc)->isObject())
    {
        c::setLastErrorMessage("doc must not be null and must be an object");
        return nullptr;
    }

    const auto& obj = toSerializableDocument(doc)->asObject();
    if (index >= obj.size())
    {
        c::setLastErrorMessage("index out of range");
        return nullptr;
    }

    return obj[index].first.c_str();
}

const dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_object_value_at(
    const dnv_vista_sdk_serializable_document_t* doc, size_t index)
{
    if (doc == nullptr || !toSerializableDocument(doc)->isObject())
    {
        c::setLastErrorMessage("doc must not be null and must be an object");
        return nullptr;
    }

    const auto& obj = toSerializableDocument(doc)->asObject();
    if (index >= obj.size())
    {
        c::setLastErrorMessage("index out of range");
        return nullptr;
    }

    return fromSerializableDocumentRef(obj[index].second);
}

const dnv_vista_sdk_serializable_document_t* dnv_vista_sdk_serializable_document_find(
    const dnv_vista_sdk_serializable_document_t* doc, const char* key)
{
    if (doc == nullptr || key == nullptr)
    {
        c::setLastErrorMessage("doc and key must not be null");
        return nullptr;
    }

    const auto* found = toSerializableDocument(doc)->find(key);
    if (found == nullptr)
    {
        return nullptr;
    }

    return fromSerializableDocumentRef(*found);
}

int dnv_vista_sdk_serializable_document_contains(const dnv_vista_sdk_serializable_document_t* doc, const char* key)
{
    if (doc == nullptr || key == nullptr)
    {
        return 0;
    }

    return toSerializableDocument(doc)->contains(key) ? 1 : 0;
}

void dnv_vista_sdk_serializable_document_set(
    dnv_vista_sdk_serializable_document_t* doc, const char* key, dnv_vista_sdk_serializable_document_t* value)
{
    if (doc == nullptr || key == nullptr || value == nullptr)
    {
        c::setLastErrorMessage("doc, key and value must not be null");
        return;
    }

    std::unique_ptr<Doc> owned{ reinterpret_cast<Doc*>(value) };
    reinterpret_cast<Doc*>(doc)->set(key, std::move(*owned));
}

void dnv_vista_sdk_serializable_document_push_back(
    dnv_vista_sdk_serializable_document_t* doc, dnv_vista_sdk_serializable_document_t* value)
{
    if (doc == nullptr || value == nullptr)
    {
        c::setLastErrorMessage("doc and value must not be null");
        return;
    }

    std::unique_ptr<Doc> owned{ reinterpret_cast<Doc*>(value) };
    reinterpret_cast<Doc*>(doc)->push_back(std::move(*owned));
}

File: bindings/rust/vista-sdk/src/transport/serializable_document.rs
use std::ffi::{CStr, CString};
use std::marker::PhantomData;
use std::ptr::NonNull;

use crate::ffi::transport::serializable_document as ffi;

/// Discriminator for the value kind held by a [`SerializableDocument`].
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum DocumentKind {
    Null,
    Boolean,
    Integer,
    Double,
    String,
    Array,
    Object,
}

/// Self-contained JSON value tree for ISO 19848 `xs:any` extension points.
///
/// Represents arbitrary JSON content (null, boolean, integer, floating point, string,
/// array, or object). Insertion order of object keys is preserved to support lossless
/// round-trip serialization.
///
/// Use [`SerializableDocumentRef`] for zero-copy read access to array elements and
/// object values borrowed from a parent document.
///
/// # Adaptation
///
/// There is no built-in dependency on any JSON library. Adapt your own JSON type by
/// implementing `From<YourJsonType> for SerializableDocument` and
/// `From<SerializableDocument> for YourJsonType` in your own crate.
pub struct SerializableDocument {
    ptr: NonNull<ffi::dnv_vista_sdk_serializable_document_t>,
}

/// Zero-copy read-only view into a [`SerializableDocument`] node.
///
/// Obtained from [`SerializableDocument::find`], [`SerializableDocument::array_at`],
/// or [`SerializableDocument::object_value_at`]. Valid only as long as the owning
/// document lives. Call [`to_owned`](SerializableDocumentRef::to_owned) to produce an
/// independent deep copy.
pub struct SerializableDocumentRef<'a> {
    ptr: *const ffi::dnv_vista_sdk_serializable_document_t,
    _marker: PhantomData<&'a SerializableDocument>,
}

unsafe impl Send for SerializableDocument {}
unsafe impl Sync for SerializableDocument {}

impl SerializableDocument {
    /// Constructs a JSON null.
    pub fn null() -> Self {
        let ptr = unsafe { ffi::dnv_vista_sdk_serializable_document_null() };
        Self {
            ptr: NonNull::new(ptr).expect("null() returned null"),
        }
    }

    /// Constructs a JSON boolean.
    pub fn from_bool(value: bool) -> Self {
        let ptr = unsafe { ffi::dnv_vista_sdk_serializable_document_from_boolean(value as i32) };
        Self {
            ptr: NonNull::new(ptr).expect("from_boolean returned null"),
        }
    }

    /// Constructs a JSON integer.
    pub fn from_i64(value: i64) -> Self {
        let ptr = unsafe { ffi::dnv_vista_sdk_serializable_document_from_integer(value) };
        Self {
            ptr: NonNull::new(ptr).expect("from_integer returned null"),
        }
    }

    /// Constructs a JSON floating point number.
    pub fn from_f64(value: f64) -> Self {
        let ptr = unsafe { ffi::dnv_vista_sdk_serializable_document_from_double(value) };
        Self {
            ptr: NonNull::new(ptr).expect("from_double returned null"),
        }
    }

    /// Constructs a JSON string.
    pub fn from_str(value: &str) -> Self {
        let c = CString::new(value).expect("value contains a NUL byte");
        let ptr = unsafe { ffi::dnv_vista_sdk_serializable_document_from_string(c.as_ptr()) };
        Self {
            ptr: NonNull::new(ptr).expect("from_string returned null"),
        }
    }

    /// Constructs an empty JSON array.
    pub fn array() -> Self {
        let ptr = unsafe { ffi::dnv_vista_sdk_serializable_document_array() };
        Self {
            ptr: NonNull::new(ptr).expect("array() returned null"),
        }
    }

    /// Constructs an empty JSON object.
    pub fn object() -> Self {
        let ptr = unsafe { ffi::dnv_vista_sdk_serializable_document_object() };
        Self {
            ptr: NonNull::new(ptr).expect("object() returned null"),
        }
    }

    /// Returns the value kind held by this document.
    pub fn kind(&self) -> DocumentKind {
        kind_from_ffi(unsafe { ffi::dnv_vista_sdk_serializable_document_kind(self.ptr.as_ptr()) })
    }

    /// Returns `true` if this document holds a JSON null.
    pub fn is_null(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_null(self.ptr.as_ptr()) }) != 0
    }

    /// Returns `true` if this document holds a JSON boolean.
    pub fn is_boolean(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_boolean(self.ptr.as_ptr()) }) != 0
    }

    /// Returns `true` if this document holds a JSON integer.
    pub fn is_integer(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_integer(self.ptr.as_ptr()) }) != 0
    }

    /// Returns `true` if this document holds a JSON floating point number.
    pub fn is_double(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_double(self.ptr.as_ptr()) }) != 0
    }

    /// Returns `true` if this document holds a JSON string.
    pub fn is_string(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_string(self.ptr.as_ptr()) }) != 0
    }

    /// Returns `true` if this document holds a JSON array.
    pub fn is_array(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_array(self.ptr.as_ptr()) }) != 0
    }

    /// Returns `true` if this document holds a JSON object.
    pub fn is_object(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_object(self.ptr.as_ptr()) }) != 0
    }

    /// Returns the boolean value.
    /// Returns `None` if this document is not a boolean.
    pub fn as_bool(&self) -> Option<bool> {
        if !self.is_boolean() {
            return None;
        }
        Some(
            (unsafe { ffi::dnv_vista_sdk_serializable_document_as_boolean(self.ptr.as_ptr()) })
                != 0,
        )
    }

    /// Returns the integer value.
    /// Returns `None` if this document is not an integer.
    pub fn as_i64(&self) -> Option<i64> {
        if !self.is_integer() {
            return None;
        }
        Some(unsafe { ffi::dnv_vista_sdk_serializable_document_as_integer(self.ptr.as_ptr()) })
    }

    /// Returns the floating point value.
    /// Returns `None` if this document is not a double.
    pub fn as_f64(&self) -> Option<f64> {
        if !self.is_double() {
            return None;
        }
        Some(unsafe { ffi::dnv_vista_sdk_serializable_document_as_double(self.ptr.as_ptr()) })
    }

    /// Returns the string value.
    /// Returns `None` if this document is not a string.
    pub fn as_str(&self) -> Option<&str> {
        if !self.is_string() {
            return None;
        }
        let raw = unsafe { ffi::dnv_vista_sdk_serializable_document_as_string(self.ptr.as_ptr()) };
        if raw.is_null() {
            return None;
        }
        Some(
            unsafe { CStr::from_ptr(raw) }
                .to_str()
                .expect("invalid UTF-8 in document string"),
        )
    }

    /// Returns the number of elements in this array, or 0 if not an array.
    pub fn array_len(&self) -> usize {
        unsafe { ffi::dnv_vista_sdk_serializable_document_array_size(self.ptr.as_ptr()) }
    }

    /// Returns a borrowed view of the array element at `index`.
    /// Returns `None` if not an array or `index` is out of range.
    pub fn array_at(&self, index: usize) -> Option<SerializableDocumentRef<'_>> {
        let raw =
            unsafe { ffi::dnv_vista_sdk_serializable_document_array_at(self.ptr.as_ptr(), index) };
        if raw.is_null() {
            None
        } else {
            Some(SerializableDocumentRef {
                ptr: raw,
                _marker: PhantomData,
            })
        }
    }

    /// Returns the number of key/value pairs in this object, or 0 if not an object.
    pub fn object_len(&self) -> usize {
        unsafe { ffi::dnv_vista_sdk_serializable_document_object_size(self.ptr.as_ptr()) }
    }

    /// Returns the key at insertion-order `index`.
    /// Returns `None` if not an object or `index` is out of range.
    pub fn object_key_at(&self, index: usize) -> Option<&str> {
        let raw = unsafe {
            ffi::dnv_vista_sdk_serializable_document_object_key_at(self.ptr.as_ptr(), index)
        };
        if raw.is_null() {
            None
        } else {
            Some(
                unsafe { CStr::from_ptr(raw) }
                    .to_str()
                    .expect("invalid UTF-8 in object key"),
            )
        }
    }

    /// Returns a borrowed view of the object value at insertion-order `index`.
    /// Returns `None` if not an object or `index` is out of range.
    pub fn object_value_at(&self, index: usize) -> Option<SerializableDocumentRef<'_>> {
        let raw = unsafe {
            ffi::dnv_vista_sdk_serializable_document_object_value_at(self.ptr.as_ptr(), index)
        };
        if raw.is_null() {
            None
        } else {
            Some(SerializableDocumentRef {
                ptr: raw,
                _marker: PhantomData,
            })
        }
    }

    /// Looks up a key in this object.
    /// Returns `None` if not an object or `key` is not present.
    pub fn find(&self, key: &str) -> Option<SerializableDocumentRef<'_>> {
        let c = CString::new(key).expect("key contains a NUL byte");
        let raw =
            unsafe { ffi::dnv_vista_sdk_serializable_document_find(self.ptr.as_ptr(), c.as_ptr()) };
        if raw.is_null() {
            None
        } else {
            Some(SerializableDocumentRef {
                ptr: raw,
                _marker: PhantomData,
            })
        }
    }

    /// Returns `true` if this is an object and contains `key`.
    pub fn contains(&self, key: &str) -> bool {
        let c = CString::new(key).expect("key contains a NUL byte");
        (unsafe {
            ffi::dnv_vista_sdk_serializable_document_contains(self.ptr.as_ptr(), c.as_ptr())
        }) != 0
    }

    /// Sets a key/value pair on this object, preserving insertion order.
    ///
    /// If this document is not yet an object it is reset to an empty object first.
    /// If `key` already exists its value is replaced in place.
    /// Takes ownership of `value`.
    pub fn set(&mut self, key: &str, value: SerializableDocument) {
        let c = CString::new(key).expect("key contains a NUL byte");
        let val_ptr = value.ptr.as_ptr();
        std::mem::forget(value);
        unsafe {
            ffi::dnv_vista_sdk_serializable_document_set(self.ptr.as_ptr(), c.as_ptr(), val_ptr)
        };
    }

    /// Appends a value to this array.
    ///
    /// If this document is not yet an array it is reset to an empty array first.
    /// Takes ownership of `value`.
    pub fn push_back(&mut self, value: SerializableDocument) {
        let val_ptr = value.ptr.as_ptr();
        std::mem::forget(value);
        unsafe { ffi::dnv_vista_sdk_serializable_document_push_back(self.ptr.as_ptr(), val_ptr) };
    }

    pub(crate) fn into_raw(self) -> *mut ffi::dnv_vista_sdk_serializable_document_t {
        let raw = self.ptr.as_ptr();
        std::mem::forget(self);
        raw
    }
}

impl Drop for SerializableDocument {
    fn drop(&mut self) {
        unsafe { ffi::dnv_vista_sdk_serializable_document_free(self.ptr.as_ptr()) };
    }
}

impl Clone for SerializableDocument {
    fn clone(&self) -> Self {
        let ptr = unsafe { ffi::dnv_vista_sdk_serializable_document_clone(self.ptr.as_ptr()) };
        Self {
            ptr: NonNull::new(ptr).expect("clone returned null"),
        }
    }
}

impl PartialEq for SerializableDocument {
    fn eq(&self, other: &Self) -> bool {
        (unsafe {
            ffi::dnv_vista_sdk_serializable_document_equals(self.ptr.as_ptr(), other.ptr.as_ptr())
        }) != 0
    }
}

impl Eq for SerializableDocument {}

impl std::fmt::Debug for SerializableDocument {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "SerializableDocument({:?})", self.kind())
    }
}

impl<'a> SerializableDocumentRef<'a> {
    pub(crate) fn from_ptr(ptr: *const ffi::dnv_vista_sdk_serializable_document_t) -> Self {
        Self {
            ptr,
            _marker: PhantomData,
        }
    }

    /// Produces an independent deep copy of this borrowed node.
    pub fn to_owned(&self) -> SerializableDocument {
        let ptr = unsafe { ffi::dnv_vista_sdk_serializable_document_clone(self.ptr) };
        SerializableDocument {
            ptr: NonNull::new(ptr).expect("clone returned null"),
        }
    }

    /// Returns the value kind of this node.
    pub fn kind(&self) -> DocumentKind {
        kind_from_ffi(unsafe { ffi::dnv_vista_sdk_serializable_document_kind(self.ptr) })
    }

    /// Returns `true` if this node holds a JSON null.
    pub fn is_null(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_null(self.ptr) }) != 0
    }

    /// Returns `true` if this node holds a JSON boolean.
    pub fn is_boolean(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_boolean(self.ptr) }) != 0
    }

    /// Returns `true` if this node holds a JSON integer.
    pub fn is_integer(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_integer(self.ptr) }) != 0
    }

    /// Returns `true` if this node holds a JSON floating point number.
    pub fn is_double(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_double(self.ptr) }) != 0
    }

    /// Returns `true` if this node holds a JSON string.
    pub fn is_string(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_string(self.ptr) }) != 0
    }

    /// Returns `true` if this node holds a JSON array.
    pub fn is_array(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_array(self.ptr) }) != 0
    }

    /// Returns `true` if this node holds a JSON object.
    pub fn is_object(&self) -> bool {
        (unsafe { ffi::dnv_vista_sdk_serializable_document_is_object(self.ptr) }) != 0
    }

    /// Returns the boolean value, or `None` if not a boolean.
    pub fn as_bool(&self) -> Option<bool> {
        if !self.is_boolean() {
            return None;
        }
        Some((unsafe { ffi::dnv_vista_sdk_serializable_document_as_boolean(self.ptr) }) != 0)
    }

    /// Returns the integer value, or `None` if not an integer.
    pub fn as_i64(&self) -> Option<i64> {
        if !self.is_integer() {
            return None;
        }
        Some(unsafe { ffi::dnv_vista_sdk_serializable_document_as_integer(self.ptr) })
    }

    /// Returns the floating point value, or `None` if not a double.
    pub fn as_f64(&self) -> Option<f64> {
        if !self.is_double() {
            return None;
        }
        Some(unsafe { ffi::dnv_vista_sdk_serializable_document_as_double(self.ptr) })
    }

    /// Returns the string value, or `None` if not a string.
    pub fn as_str(&self) -> Option<&str> {
        if !self.is_string() {
            return None;
        }
        let raw = unsafe { ffi::dnv_vista_sdk_serializable_document_as_string(self.ptr) };
        if raw.is_null() {
            return None;
        }
        Some(
            unsafe { CStr::from_ptr(raw) }
                .to_str()
                .expect("invalid UTF-8 in document string"),
        )
    }

    /// Returns the number of elements in this array, or 0 if not an array.
    pub fn array_len(&self) -> usize {
        unsafe { ffi::dnv_vista_sdk_serializable_document_array_size(self.ptr) }
    }

    /// Returns a borrowed view of the array element at `index`.
    /// Returns `None` if not an array or `index` is out of range.
    pub fn array_at(&self, index: usize) -> Option<SerializableDocumentRef<'a>> {
        let raw = unsafe { ffi::dnv_vista_sdk_serializable_document_array_at(self.ptr, index) };
        if raw.is_null() {
            None
        } else {
            Some(SerializableDocumentRef {
                ptr: raw,
                _marker: PhantomData,
            })
        }
    }

    /// Returns the number of key/value pairs in this object, or 0 if not an object.
    pub fn object_len(&self) -> usize {
        unsafe { ffi::dnv_vista_sdk_serializable_document_object_size(self.ptr) }
    }

    /// Returns the key at insertion-order `index`, or `None` if out of range.
    pub fn object_key_at(&self, index: usize) -> Option<&str> {
        let raw =
            unsafe { ffi::dnv_vista_sdk_serializable_document_object_key_at(self.ptr, index) };
        if raw.is_null() {
            None
        } else {
            Some(
                unsafe { CStr::from_ptr(raw) }
                    .to_str()
                    .expect("invalid UTF-8 in object key"),
            )
        }
    }

    /// Returns a borrowed view of the object value at insertion-order `index`.
    /// Returns `None` if not an object or `index` is out of range.
    pub fn object_value_at(&self, index: usize) -> Option<SerializableDocumentRef<'a>> {
        let raw =
            unsafe { ffi::dnv_vista_sdk_serializable_document_object_value_at(self.ptr, index) };
        if raw.is_null() {
            None
        } else {
            Some(SerializableDocumentRef {
                ptr: raw,
                _marker: PhantomData,
            })
        }
    }

    /// Looks up a key in this object.
    /// Returns `None` if not an object or `key` is not present.
    pub fn find(&self, key: &str) -> Option<SerializableDocumentRef<'a>> {
        let c = CString::new(key).expect("key contains a NUL byte");
        let raw = unsafe { ffi::dnv_vista_sdk_serializable_document_find(self.ptr, c.as_ptr()) };
        if raw.is_null() {
            None
        } else {
            Some(SerializableDocumentRef {
                ptr: raw,
                _marker: PhantomData,
            })
        }
    }

    /// Returns `true` if this is an object and contains `key`.
    pub fn contains(&self, key: &str) -> bool {
        let c = CString::new(key).expect("key contains a NUL byte");
        (unsafe { ffi::dnv_vista_sdk_serializable_document_contains(self.ptr, c.as_ptr()) }) != 0
    }
}

impl<'a> std::fmt::Debug for SerializableDocumentRef<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "SerializableDocumentRef({:?})", self.kind())
    }
}

fn kind_from_ffi(k: ffi::dnv_vista_sdk_serializable_document_kind_t) -> DocumentKind {
    match k {
        ffi::dnv_vista_sdk_serializable_document_kind_t::Null => DocumentKind::Null,
        ffi::dnv_vista_sdk_serializable_document_kind_t::Boolean => DocumentKind::Boolean,
        ffi::dnv_vista_sdk_serializable_document_kind_t::Integer => DocumentKind::Integer,
        ffi::dnv_vista_sdk_serializable_document_kind_t::Double => DocumentKind::Double,
        ffi::dnv_vista_sdk_serializable_document_kind_t::String => DocumentKind::String,
        ffi::dnv_vista_sdk_serializable_document_kind_t::Array => DocumentKind::Array,
        ffi::dnv_vista_sdk_serializable_document_kind_t::Object => DocumentKind::Object,
    }
}
